user_tools/src/spark_rapids_tools/tools/qualx/preprocess.py

```diff
@@ -596,6 +599,16 @@ def combine_tables(table_name: str) -> pd.DataFrame:
     app_runtype = toc[['appId', 'runType']].drop_duplicates()
     full_tbl = full_tbl.merge(app_runtype, on='appId')
 
+    # add duration_ratio feature
+    full_tbl['duration_ratio'] = full_tbl['duration_sum'] / full_tbl['Duration']
+    full_tbl['duration_ratio'] = full_tbl['duration_ratio'].replace([np.inf, -np.inf], 0).fillna(0)
+
+    # add failed tasks features
+    failed_tasks_tbl = combine_tables('failed_tasks_tbl')
+    full_tbl = full_tbl.merge(failed_tasks_tbl, on=['appName', 'appId', 'sqlID'], how='left')
+    full_tbl['failed_tasks_ratio'] = full_tbl['failed_tasks'] / full_tbl['numTasks_sum']
+    full_tbl.fillna({'failed_tasks': 0, 'failed_tasks_ratio': 0.0}, inplace=True)
+
```

Comment:
My biggest concern is that there is an old code in scala module that does something similar.
This implies that we have two different components that apply some heuristics to to calculate the durations.
The paradox is that we have to apply some logic in the scala to calculate the durations anyway. 

Do we still want to have that duration feature here until we revisit the duration calculations in the Scala side?

Reply:
I think the only durations being used here is actually just the ratio of the existing `duration_sum` to `Duration` columns.  Both of these are obtained from the Profiler CSV files, so we're only computing the ratio of the two.  If the scala code is _also_ generating a ratio, then we would just use the output of that.

Reply:
That said, I imagine that the scala code provides the basic/raw metrics, while qualx essentially doing "feature engineering" on those raw metrics (in this case, generating a ratio of the raw metrics which is trying to capture some representation of the ratio of task time to wall clock).